firedrake/paraview_reordering.py

```diff
@@ -0,0 +1,449 @@
+from tsfc.fiatinterface import create_element
+import numpy as np
+
+
+def firedrake_local_to_cart(element):
+    r"""Gets the list of nodes for an element (provided they exist.)
+    :arg element: a ufl element.
+    :returns: a list of arrays of floats where each array is a node.
+    """
+    fiat_element = create_element(element, vector_is_mixed=False)
+    # TODO: Surely there is an easier way that I've forgotten?
+    carts = [np.array(list(phi.get_point_dict().keys())[0])
+             for phi in fiat_element.dual_basis()]
+    return carts
+
+
+def invert(list1, list2):
+    r"""Given two maps (lists) from [0..N] to nodes, finds a permutations between them.
+    :arg list1: a list of nodes.
+    :arg list2: a second list of nodes.
+    :returns a list of integers, l, such that list1[x] = list2[l[x]]
+    """
+    if len(list1) != len(list2):
+        raise Exception("Dimension of Paraview basis and Element basis unequal.")
+
+    def find_same(val, lst, tol=0.0000001):
+        for (idx, x)in enumerate(lst):
+            if np.linalg.norm(val - x) < tol:
+                return idx
+        raise Exception("Basis of element incompatible with paraview basis: node {0} is not in {1}".format(val, lst))
+    # finds map from idx of list 1 to list of list 2.
+    perm = [find_same(x, list2) for x in list1]
+    #  check for bijectivity:
+    if len(set(perm)) != len(list1):
+        raise Exception("Unable to establish permutation between Paraview basis and given element basis.")
```

Comment:
```python
N = len(list1)
pt1, o1 = zip(*sorted(zip(map(tuple, list1), range(N))))
pt2, o2 = zip(*sorted(zip(map(tuple, list2), range(N))))
if not np.allclose(pt1, pt2):
    raise ValueError(...)

return numpy.asarray(o2)[numpy.argsort(o2)
```

NlogN, rather than N^3. And you get to feel like you were writing Haskell.